internal/pkg/image/packer/gocryptfs.go

```diff
@@ -0,0 +1,214 @@
+package packer
+
+import (
+	"bytes"
+	"crypto/md5"
+	"crypto/rand"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"os/signal"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"syscall"
+
+	"github.com/apptainer/apptainer/internal/pkg/image/driver"
+	"github.com/apptainer/apptainer/internal/pkg/util/bin"
+	"github.com/apptainer/apptainer/pkg/sylog"
+)
+
+type Gocryptfs struct {
+	*Squashfs
+	gocryptfsPath string
+	Pass          string
+}
+
+type cryptInfo struct {
+	cipherDir, plainDir, pass, confPath, parentDir string
+	pid                                            int
+}
+
+func newCryptInfo() *cryptInfo {
+	return &cryptInfo{
+		cipherDir: "",
+		plainDir:  "",
+		pass:      "",
+		confPath:  "",
+		parentDir: "",
+		pid:       -1,
+	}
+}
+
+func NewGocryptfs() *Gocryptfs {
+	g := &Gocryptfs{
+		Squashfs: NewSquashfs(),
+	}
+	g.gocryptfsPath, _ = bin.FindBin("gocryptfs")
+	return g
+}
+
+func (g *Gocryptfs) HasGocryptfs() bool {
+	return g.gocryptfsPath != ""
+}
+
+func (g *Gocryptfs) init(tmpDir string) (cryptInfo *cryptInfo, err error) {
+	if !g.HasGocryptfs() {
+		return nil, fmt.Errorf("either gocryptfs or fusermount does not exist")
+	}
+
+	cryptInfo = newCryptInfo()
+	parentDir, err := os.MkdirTemp(tmpDir, "gocryptfs-")
+	if err != nil {
+		return
+	}
+	cryptInfo.parentDir = parentDir
+	cipherDir := filepath.Join(parentDir, "cipher")
+	plainDir := filepath.Join(parentDir, "plain")
+
+	err = os.Mkdir(cipherDir, 0o700)
+	if err != nil {
+		return
+	}
+	cryptInfo.cipherDir = cipherDir
+	err = os.Mkdir(plainDir, 0o700)
+	if err != nil {
+		return
+	}
+	cryptInfo.plainDir = plainDir
+
+	buf := make([]byte, 32)
+	_, err = rand.Read(buf)
+	if err != nil {
+		return
+	}
+
+	hash := md5.Sum(buf)
+	cryptInfo.pass = hex.EncodeToString(hash[:])
+	pass := fmt.Sprintf("%s\n", cryptInfo.pass)
+	sylog.Debugf("Start initializing gocryptfs, cipher: %s, plain: %s", cipherDir, plainDir)
+	cmd := exec.Command(g.gocryptfsPath, "-init", "-deterministic-names", "-plaintextnames", cipherDir)
+	cmd.Stdin = strings.NewReader(pass + pass)
+	if err = cmd.Run(); err != nil {
+		return
+	}
+	cryptInfo.confPath = filepath.Join(cipherDir, "gocryptfs.conf")
+
+	cmd = exec.Command(g.gocryptfsPath, cipherDir, plainDir)
+	cmd.Stdin = strings.NewReader(pass)
+	if err = cmd.Run(); err != nil {
+		return
+	}
+	realPid, err := getGocryptfsPid(fmt.Sprintf("-notifypid=%d", cmd.Process.Pid))
+	if err != nil {
+		return
+	}
+	cryptInfo.pid = realPid
+	// trap SIGINT/SIGTERM signals
+	go func() {
+		ch := make(chan os.Signal, 1)
+		signal.Notify(ch, syscall.SIGINT, syscall.SIGTERM)
+		<-ch
+		sylog.Warningf("Received SIGINT/SIGTERM signal")
+		driver.KillProcess(cmd.Process.Pid, true, "gocryptfs")
+	}()
```

Comment:
I take it from this code that it wasn't possible to directly use the image driver here.  What prevents that?

Reply:
InitImageDrivers does not require engine configuration, it only requires an apptainerconf.File which is a representation of apptainer.conf.  It's not hard to get.

It doesn't look to me like anything is needed from a session.  This internal image driver doesn't even appear to use anything from the DriverParams structure; you could probably just pass nil.  It does use image.MountParams but it shouldn't be difficult to create one of those structures to pass in for the parameters it needs.